packages/@react-aria/menu/src/useSubMenuTrigger.ts

```diff
@@ -0,0 +1,244 @@
+          e.stopPropagation();
+          onSubMenuClose();
+        }
+        break;
+      case 'ArrowRight':
+        if (direction === 'rtl') {
+          e.stopPropagation();
+          onSubMenuClose();
+        }
+        break;
+      case 'Escape':
+        e.stopPropagation();
+        state.closeAll();
+        break;
+    }
+  };
+
+  let subMenuProps = {
+    id: overlayId,
+    'aria-label': node.textValue,
+    ...(subMenuType === 'menu' && {
+      onClose: state.closeAll,
+      autoFocus: state.focusStrategy,
+      onKeyDown: subMenuKeyDown
+    })
+  };
+
+  let subMenuTriggerKeyDown = (e: KeyboardEvent) => {
+    switch (e.key) {
+      case 'ArrowRight':
+        if (!isDisabled) {
+          if (direction === 'ltr') {
+            if (subMenuType === 'menu' && !!subMenuRef?.current && document.activeElement === ref?.current) {
+              subMenuRef.current.focus();
+            } else {
+              onSubmenuOpen('first');
+            }
+          } else if (state.isOpen) {
+            e.stopPropagation();
+            onSubMenuClose();
+          }
+        }
+
+        break;
+      case 'ArrowLeft':
+        if (!isDisabled) {
+          if (direction === 'rtl') {
+            if (subMenuType === 'menu' && !!subMenuRef?.current && document.activeElement === ref?.current) {
+              subMenuRef.current.focus();
+            } else {
+              onSubmenuOpen('first');
+            }
+          } else if (state.isOpen) {
+            e.stopPropagation();
+            onSubMenuClose();
+          }
+        }
+        break;
+    }
+  };
+
+  let onPressStart = (e: PressEvent) => {
+    if (!isDisabled && (e.pointerType === 'virtual' || e.pointerType === 'keyboard')) {
+      // If opened with a screen reader or keyboard, auto focus the first submenu item.
+      onSubmenuOpen('first');
+    }
+  };
+
+  let onPress = (e: PressEvent) => {
+    if (!isDisabled && e.pointerType === 'touch') {
+      onSubmenuOpen();
+    }
+  };
+
+  let onHoverChange = (isHovered) => {
+    if (!isDisabled) {
+      if (isHovered && !state.isOpen) {
+        if (!openTimeout.current) {
+          openTimeout.current = setTimeout(() => {
+            onSubmenuOpen();
+          }, 200);
+        }
+      } else if (!isHovered) {
+        cancelOpenTimeout();
+      }
+    }
+  };
+
+  let onBlur = (e) => {
+    // TODO: getting rid of the (!isElementInChildOfActiveScope(e.relatedTarget) part of the check below fixes hovering to open a root unavailable menu item when focus is on a submenu
+    // but breaks the case where the user is hovering over a submenu's submenu trigger child item and then hovers a root menu item (eg hover lvl 1 item 2 -> hover lvl 2 item 3 -> hover lvl 1 item 3)
+    // Ideally we'd be able to track the full menu tree and then check if focus has moved to an element that isn't part of the current submenu tree and close it then
+    if (state.isOpen && parentMenuRef.current.contains(e.relatedTarget)) {
```

Comment:
captured in ticket, will want to eventually be able to track if any arbitrary click/focus movement is within a root menu's tree or outside the menu tree

Reply:
https://github.com/orgs/adobe/projects/19/views/4?pane=issue&itemId=39885795